pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/server/ServerPlanRequestUtils.java

```diff
@@ -291,4 +169,106 @@ private static void attachTimeBoundary(PinotQuery pinotQuery, TimeBoundaryInfo t
       pinotQuery.setFilterExpression(timeFilterExpression);
     }
   }
+
+  /**
+   * attach the dynamic filter to the given PinotQuery.
+   */
+  static void attachDynamicFilter(PinotQuery pinotQuery, JoinNode.JoinKeys joinKeys, List<Object[]> dataContainer,
+      DataSchema dataSchema) {
```

Comment:
TODO: handle 
1. empty dataContainer,
2. empty existing filter
3. null dataContainer or errored dataContainer

Reply:
[update]
- 1 & 2 handled
- 3 should be handled by error handling on the pipeline context level